File: DynamicPrgramming/91. Decode Ways/Solution.py
class Solution(object):
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s);
        if n==0:
            return 0
        f=[0 for i in range(n+1)]
        f[0]=1
        for i in range(1,n+1):
            if ord(s[i-1])>ord('0'):
                f[i]=f[i-1]
            if i>1 and (ord(s[i-2])==ord('1') or (ord(s[i-2])==ord('2') and ord(s[i-1])<=ord('6'))):
                f[i]+=f[i-2]
        return f[n]
        

File: DynamicPrgramming/91. Decode Ways/Solution.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        if (n==0){
            return 0;
        }
        int[] f = new int[n+1];
        f[0]=1;
        for (int i=1; i<=n; i++){
            if (s.charAt(i-1)>'0'){
                f[i] = f[i-1];
            }
            if (i>1){
                if (s.charAt(i-2)=='1' || s.charAt(i-2)=='2'&&s.charAt(i-1)<='6'){
                    f[i]+=f[i-2];
                }
            }
        }
        return f[n];
    }
}